text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,375 @@
+
+**Conclusion**: `bazel` shows that a hash-based workflow seems to work well enough, making an argument for the use of it in `cargo` too. It also uses the current user, to prevent attacks by having compiled a program as root and making the directory accessible to other users later on by also compiling there for them. `cargo` could also do this, though I do not know what happens when `--output_user_root` is set to the same path for two different users.
+
+*Note: I looked at Bazel 5.4.0, the latest stable version as of this writing, things may change in the future or be different for older versions.*
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+-->
+
+- Do we want to differentiate according to users ? `bazel` is a generic build tool, whereas `cargo` is not, so maybe differentiating on users is not necessary for us ?
+
+## Handle possibly thousands of directories in a single `CARGO_TARGET_BASE_DIR` path
+
+While a single dev machine is unlikely to have enough projects that the simple scheme of `<project name>-<hash>` will produce enough directories to slow down working in `$CARGO_TARGET_BASE_DIR/`, it could still happen, and notably in private CI, which are often less compartimentalized than public ones. Simple cruft over time (i.e, never calling `cargo clean` over years) could also make it happen, if much slower.
+
+A quick design that would probably work and still be relatively navigable by humans is: instead of `project-name-<hash>/`, use `p/r/o/j/e/ct-name-<hash>/`. Here I used an arbitrary limit of 5 subdirectories but really, anything could be used. A problem occurs if a `.` is present in the directory name,but in that case we could simply append it to the symbol before like so: `proj....name/` -> `p/r/o/j..../n/ame-<hash>/`.
+
+If the project name contains a space in the first 5 symbols, maybe replace it like so: `proj name/` -> `p/r/o/j-space/n/ame-<hash>/`.
+
+I don't know of any filesystem that has troubles with `_` or `-` (often used as delimiters) so those have no special handling.
+
+If, on the other, we don't care (or care less) for human usability of the directories inside `CARGO_TARGET_BASE_DIR`, we could simply invert the format to use `<hash>-<project name>`: since the hash would have fixed length and set of characters, it would be easy to split into subdirectories without special casing and a simple `find . -n 'project name'` would point to the final target directory for the project (or all of them, if several similarly named projects exists on the machine).
```

Comment:
@estebank Here I discuss how to handle scaling.

I like the final idea I proposed but I'm waiting for other opinions before changing this RFC-wide

Reply:
Something that will help explode the number of directories is ["cargo script"](https://github.com/rust-lang/cargo/issues/12207).  Originally, my plan was to use a shared CARGO_TARGET_DIR as a build-time optimization but with some recent discussions on per-user caching, I'm hopeful we can instead have per-script CARGO_TARGET_DIR. 

The current implementation is sort-of something like an unconditional `CARGO_TARGET_BASE_DIR=$CARGO_HOME/target` where the intermediate directories are `{hash[0:2]}/hash[2:]}` where the hash is just a "short hash".  Sharding the directory by 255 should be sufficient for any system....

Reply:
- https://github.com/rust-lang/cargo/blob/master/src/cargo/core/workspace.rs#L387
- https://github.com/rust-lang/cargo/blob/master/src/cargo/util/hex.rs#L29

Reply:
underspecified naming scheme to allow for this